src/App.tsx

```diff
@@ -1,24 +1,252 @@
+
+  const completedTodoIds = useMemo(() => completedTodos
+    .map(todo => todo.id), [completedTodos]);
+
+  const changeTodosIdsToRemove = useCallback((value: number) => {
+    setTodosIdsToRemove((currentIds) => [...currentIds, value]);
+  }, []);
+
+  const deleteCompletedTodos = useCallback(() => {
+    completedTodoIds.map(id => changeTodosIdsToRemove(id));
+
+    completedTodoIds.map(id => deleteTodo(id));
+  }, [completedTodoIds]);
+
+  const removeDeleteId = (removedId: number) => {
+    setTodosIdsToRemove((currentIds) => currentIds
+      .filter(id => id !== removedId));
+  };
+
+  const changeTodosIdsToUpdate = (value: number) => {
+    setTodosIdsToUpdate((currentIds) => [...currentIds, value]);
+  };
+
+  const removeUpdatedId = (idToRemove: number) => {
+    setTodosIdsToUpdate((currentIds) => currentIds
+      .filter(id => id !== idToRemove));
+  };
+
+  const resetTodosIdsToUpdate = () => {
+    setTodosIdsToUpdate([]);
+  };
+
+  const toggleAllTodosStatus = useCallback(async (status: boolean) => {
+    todos.forEach(todo => {
+      if (todo.completed !== status) {
+        changeTodosIdsToUpdate(todo.id);
+      }
+    });
+
+    await Promise.all(todos
+      .map(todo => changeCompletedStatus(todo.id, status)));
+
+    resetTodosIdsToUpdate();
+  }, [todos]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <NewFormTodo
+          todoTitle={todoTitle}
+          onTodoTitleChange={onTodoTitleChange}
+          handleErrors={handleErrors}
+          addNewTodo={addNewTodo}
+          isLoading={isAddWaiting}
+          isAllTodosCompleted={isAllTodosCompleted}
+          haveTodos={haveTodos}
+          toggleAllTodosStatus={toggleAllTodosStatus}
+        />
+
+        {(todos.length > 0 || tempTodo !== null) && (
+          <>
+            <TodoList
+              todos={visibleTodos}
+              tempTodo={tempTodo}
+              deleteTodo={deleteTodo}
+              isDeleteWaiting={isDeleteWaiting}
+              todosIdsToRemove={todosIdsToRemove}
+              changeTodosIdsToRemove={changeTodosIdsToRemove}
+              removeDeleteId={removeDeleteId}
+              changeCompletedStatus={changeCompletedStatus}
+              isUpdateWaiting={isUpdateWaiting}
+              todosIdsToUpdate={todosIdsToUpdate}
+              changeTodosIdsToUpdate={changeTodosIdsToUpdate}
+              removeUpdatedId={removeUpdatedId}
+              changeTodoTitle={changeTodoTitle}
+            />
```

Comment:
check the code, it's so many props that you pass to the child component
can we decrease the number, and move some functions and properties from main component to the child ones?